opm/simulators/wells/WellInterface_impl.hpp

```diff
@@ -488,6 +488,36 @@ namespace Opm
             this->operability_status_.solvable = true;
             bool converged = this->iterateWellEquations(ebosSimulator, dt, well_state, group_state, deferred_logger);
 
+            // Newly opened wells with THP control sometimes struggles to
+            // converge due to bad initial guess. Or due to the simple fact
+            // that the well needs to change to another control.
+            // We therefore try to solve the well with BHP control to get
+            // an better initial guess.
+            // If the well is supposed to operate under THP control
+            // "updateWellControl" will switch it back to THP later.
+            if (!converged && this->changed_to_open_this_step_) {
+                auto& ws = well_state.well(this->indexOfWell());
+                bool thp_control = false;
+                if (this->well_ecl_.isInjector()) {
+                    thp_control = ws.injection_cmode == Well::InjectorCMode::THP;
+                    if (thp_control) {
+                        ws.injection_cmode = Well::InjectorCMode::BHP;
+                        this->well_control_log_.push_back(Well::InjectorCMode2String(Well::InjectorCMode::THP));
+                    }
+                } else {
+                    thp_control = ws.production_cmode == Well::ProducerCMode::THP;
+                    if (thp_control) {
+                        ws.production_cmode = Well::ProducerCMode::BHP;
+                        this->well_control_log_.push_back(Well::ProducerCMode2String(Well::ProducerCMode::THP));
+                    }
+                }
+                if (thp_control) {
+                    const std::string msg = std::string("The newly opened well ") + this->name()
+                        + std::string(" with THP control did not converge during inner iterations, we try again with bhp control");
+                    deferred_logger.debug(msg);
+                    converged = this->iterateWellEquations(ebosSimulator, dt, well_state, group_state, deferred_logger);
+                }
+            }
```

Comment:
Good idea. I just moved the solve to the "solveWellEquation" that is called before updateWellControl for every iteration. I have verified that it still helps for my problematic case.

Reply:
Thanks for the update. It is still different from what I am suggesting. But as I said, I understand why this works for some scenarios and importantly, it fixes your targeting scenario. 

Please check the results for the jenkins running and make sure it did not break anything in an undesirable way.  And address the small comments above.  Then it can go in.